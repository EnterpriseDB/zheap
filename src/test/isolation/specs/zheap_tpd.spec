# Simple tests that show the working of zheap correctly with TPD.
setup
{
 CREATE TABLE animals (name text, counter int) WITH (storage_engine = 'zheap');
 INSERT INTO animals VALUES ('cat', 1), ('dog', 10), ('monkey', 100);
}

teardown
{
 DROP TABLE animals;
}

session "s1"
setup		{ BEGIN; }
step "r1"	{ SELECT * FROM animals; }
step "i1"	{ INSERT INTO ANIMALS VALUES ('cow', 11); }
step "t1"	{ DELETE FROM ANIMALS WHERE name = 'dog'; }
step "c1"	{ COMMIT; }

# insert and index key update
session "s2"
setup       { BEGIN; }
step "w2"	{ UPDATE animals SET counter = counter + 2 WHERE name = 'cat'; }
step "i2"	{ INSERT INTO ANIMALS VALUES ('lion', 22); }
step "r2"	{ SELECT * FROM animals; }
step "t2"	{ DELETE FROM ANIMALS WHERE name = 'lion'; }
step "c2"	{ COMMIT; }
step "d2"	{ ROLLBACK; }

# tuple size increase
session "s3"
setup       { BEGIN; }
step "i3"	{ INSERT INTO ANIMALS VALUES ('panther', 33); }
step "w3"	{ UPDATE animals SET counter = counter + 3 WHERE name = 'cat'; }
step "t3"	{ DELETE FROM ANIMALS WHERE counter < 3; }
step "r3"	{ SELECT * FROM animals; }
step "c3"	{ COMMIT; }

# index key update
session "s4"
setup       { BEGIN; }
step "i4"	{ INSERT INTO ANIMALS VALUES ('giraffe', 44); }
step "w4"	{ UPDATE animals SET counter = counter + 4 WHERE name = 'cat'; }
step "r4"	{ SELECT * FROM animals; }
step "c4"	{ COMMIT; }

# insert and index key update
session "s5"
setup       { BEGIN; }
step "i5"	{ INSERT INTO ANIMALS VALUES('tiger', 55); }
step "w5"	{ UPDATE animals SET counter = counter + 5 WHERE name = 'cat'; }
step "r5"	{ SELECT * FROM animals; }
step "t5"	{ DELETE FROM ANIMALS WHERE name = 'dog'; }
step "c5"	{ COMMIT; }

# check if the insertions are handled with TPD pages
permutation "i1" "i2" "i3" "i4" "i5" "c5" "r5" "c4" "r4" "c2" "r2" "c3" "r3" "c1" "r1"

# check if the updates with rollback are handled are with TPD pages
permutation "i1" "i2" "i3" "i4" "i5" "w5" "c5" "r5" "w2" "d2" "r2" "w3" "c3" "r3" "c1" "c4"

# check for the correct handling of deletes with rollback with TPD pages
permutation "i1" "i2" "i3" "i4" "i5" "t5" "c5" "r5" "t2" "d2" "r2" "t3" "c3" "r3" "c1" "c4"
