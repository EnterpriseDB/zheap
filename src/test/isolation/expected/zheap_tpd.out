Parsed test spec with 5 sessions

starting permutation: i1 i2 i3 i4 i5 c5 r5 c4 r4 c2 r2 c3 r3 c1 r1
step i1: INSERT INTO ANIMALS VALUES ('cow', 11);
step i2: INSERT INTO ANIMALS VALUES ('lion', 22);
step i3: INSERT INTO ANIMALS VALUES ('panther', 33);
step i4: INSERT INTO ANIMALS VALUES ('giraffe', 44);
step i5: INSERT INTO ANIMALS VALUES('tiger', 55);
step c5: COMMIT;
step r5: SELECT * FROM animals;
name           counter        

cat            1              
dog            10             
monkey         100            
tiger          55             
step c4: COMMIT;
step r4: SELECT * FROM animals;
name           counter        

cat            1              
dog            10             
monkey         100            
giraffe        44             
tiger          55             
step c2: COMMIT;
step r2: SELECT * FROM animals;
name           counter        

cat            1              
dog            10             
monkey         100            
lion           22             
giraffe        44             
tiger          55             
step c3: COMMIT;
step r3: SELECT * FROM animals;
name           counter        

cat            1              
dog            10             
monkey         100            
lion           22             
panther        33             
giraffe        44             
tiger          55             
step c1: COMMIT;
step r1: SELECT * FROM animals;
name           counter        

cat            1              
dog            10             
monkey         100            
cow            11             
lion           22             
panther        33             
giraffe        44             
tiger          55             

starting permutation: i1 i2 i3 i4 i5 w5 c5 r5 w2 d2 r2 w3 c3 r3 c1 c4
step i1: INSERT INTO ANIMALS VALUES ('cow', 11);
step i2: INSERT INTO ANIMALS VALUES ('lion', 22);
step i3: INSERT INTO ANIMALS VALUES ('panther', 33);
step i4: INSERT INTO ANIMALS VALUES ('giraffe', 44);
step i5: INSERT INTO ANIMALS VALUES('tiger', 55);
step w5: UPDATE animals SET counter = counter + 5 WHERE name = 'cat';
step c5: COMMIT;
step r5: SELECT * FROM animals;
name           counter        

cat            6              
dog            10             
monkey         100            
tiger          55             
step w2: UPDATE animals SET counter = counter + 2 WHERE name = 'cat';
step d2: ROLLBACK;
step r2: SELECT * FROM animals;
name           counter        

cat            6              
dog            10             
monkey         100            
tiger          55             
step w3: UPDATE animals SET counter = counter + 3 WHERE name = 'cat';
step c3: COMMIT;
step r3: SELECT * FROM animals;
name           counter        

cat            9              
dog            10             
monkey         100            
panther        33             
tiger          55             
step c1: COMMIT;
step c4: COMMIT;

starting permutation: i1 i2 i3 i4 i5 t5 c5 r5 t2 d2 r2 t3 c3 r3 c1 c4
step i1: INSERT INTO ANIMALS VALUES ('cow', 11);
step i2: INSERT INTO ANIMALS VALUES ('lion', 22);
step i3: INSERT INTO ANIMALS VALUES ('panther', 33);
step i4: INSERT INTO ANIMALS VALUES ('giraffe', 44);
step i5: INSERT INTO ANIMALS VALUES('tiger', 55);
step t5: DELETE FROM ANIMALS WHERE name = 'dog';
step c5: COMMIT;
step r5: SELECT * FROM animals;
name           counter        

cat            1              
monkey         100            
tiger          55             
step t2: DELETE FROM ANIMALS WHERE name = 'lion';
step d2: ROLLBACK;
step r2: SELECT * FROM animals;
name           counter        

cat            1              
monkey         100            
tiger          55             
step t3: DELETE FROM ANIMALS WHERE counter < 3;
step c3: COMMIT;
step r3: SELECT * FROM animals;
name           counter        

monkey         100            
panther        33             
tiger          55             
step c1: COMMIT;
step c4: COMMIT;
